### PR TITLE
DS VMware: Fix ipv6 addr converter from netinfo to netifaces

### PR DESCRIPTION
**Pull request by PengpengSun**
## Proposed Commit Message

```
DS VMware: Fix ipv6 addr converter from netinfo to netifaces

Found an issue when verifying PR 4634 on vSphere platform, 
which is failing to convert ipv6 addr from netinfo format to
netifaces format due to 'bcast' is not existing in ipv6.

This PR is fixing this by updating ipv4 converter function and
adding a new ipv6 converter function,  also adding unit tests.
```

## Checklist

- [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/development/index.html)
- [x] I have updated or added any [unit tests](https://cloudinit.readthedocs.io/en/latest/development/testing.html) accordingly
- [x] I have updated or added any [documentation](https://cloudinit.readthedocs.io/en/latest/development/contribute_docs.html) accordingly

## Merge type

- [x] Squash merge using "Proposed Commit Message"
- [ ] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

### CODE REVIEW

**Comment by PengpengSun** [comment]
Hi @catmsred,
FYI, and it will be great if you could help to review this change.

File: cloudinit/sources/DataSourceVMware.py [hunk 1]
```diff
@@ -821,38 +821,73 @@ def is_valid_ip_addr(val):
...
+    {
+      "ip": "2001:db8:abcd:1234::1/64",
+      "scope6": "global",
     }
+    """
+    addr_in_netifaces_format = {}
+    if "ip" in addr:
+        ipv6_with_netmask = addr["ip"]
+        ipv6_parts = ipv6_with_netmask.split("/")
+        ipv6_address = ipv6_parts[0]
```
Comment (catmsred) [inline cloudinit/sources/DataSourceVMware.py:873]: Is there any chance this will ever process anything without a / in it?  Would it make sense put these dereferences in a try/except group?

File: tests/unittests/sources/test_vmware.py [hunk 2]
```diff
@@ -147,6 +156,18 @@ def test_no_data_access_method(self):
             ret = ds.get_data()
         self.assertFalse(ret)
 
+    def test_convert_to_netifaces_ipv4_format(self):
```
Comment (catmsred) [inline tests/unittests/sources/test_vmware.py:159]: Are there any failure or edge cases that should be considered/tested?

**Review by catmsred — APPROVED** [review]
Thanks for cleaning up the ipv6 address handling in this function!  I had a couple questions but nothing that needs to be blocking.

**Comment by holmanb** [comment]
> Found an issue when verifying PR 4634 on vSphere platform, 
which is failing to convert ipv6 addr from netinfo format to
netifaces format due to 'bcast' is not existing in ipv6.

@PengpengSun could you please provide cloud-init logs or at least a the traceback of the error?

File: cloudinit/sources/DataSourceVMware.py (hunk 1 above)
Comment (holmanb) [reply cloudinit/sources/DataSourceVMware.py:873]: Good catch.

It looks like when an interface is [point-to-point this is possible](https://github.com/canonical/cloud-init/blob/9c77b3896c903cadd966c71b7301593c06145520/cloudinit/netinfo.py#L89).

File: cloudinit/sources/DataSourceVMware.py [hunk 3]
```diff
@@ -821,38 +821,73 @@ def is_valid_ip_addr(val):
...
+      "netmask": "ffff:ffff:ffff:ffff::/64",
+      "addr": "2001:db8:abcd:1234::1"
+    }
+    cloudinit.netinfo ipv6 format:
+    {
+      "ip": "2001:db8:abcd:1234::1/64",
+      "scope6": "global",
     }
+    """
+    addr_in_netifaces_format = {}
```
Comment (holmanb) [inline cloudinit/sources/DataSourceVMware.py:869]: This function does manual string parsing, which shouldn't be required due to Python standard library code.

Maybe you could use something like this? (I haven't tested it)

```python
def convert_to_netifaces_ipv6_format(addr: dict) -> dict:
    """
    Takes a cloudinit.netinfo formatted address and converts to netifaces
    format, since this module was originally written with netifaces as the
    network introspection module.
    netifaces ipv6 format:
    {
      "netmask": "ffff:ffff:ffff:ffff::/64",
      "addr": "2001:db8:abcd:1234::1"
    }
    cloudinit.netinfo ipv6 format:
    {
      "ip": "2001:db8:abcd:1234::1/64",
      "scope6": "global",
    }
    """
    if not addr.get("ip"):
        return {}
    ipv6 = ipaddress.IPv6Interface(addr.get("ip"))
    return {
        "netmask": f"{ipv6.netmask}/{ipv6.network.prefixlen}",
        "addr": str(ipv6.ip)
    }
```

**Review by holmanb — CHANGES_REQUESTED** [review]

**Comment by PengpengSun** [comment]
> > Found an issue when verifying PR 4634 on vSphere platform,
> > which is failing to convert ipv6 addr from netinfo format to
> > netifaces format due to 'bcast' is not existing in ipv6.
> 
> @PengpengSun could you please provide cloud-init logs or at least a traceback of the error?

Sure, please find the traceback as below:

2024-03-07 02:53:41,668 - DataSourceVMware.py[DEBUG]: device ens224 has more than one ipv6 address: [{'ip': '2620:124:6020:c716:0:a:0:7a5/128', 'scope6': 'global'}, {'ip': '2620:124:6020:c716:250:56ff:fe87:41f8/64', 'scope6': 'global'}, {'ip': 'fd00:1234:5678:9abc::2/64', 'scope6': 'global'}, {'ip': 'fe80::250:56ff:fe87:41f8/64', 'scope6': 'link'}]
2024-03-07 02:53:41,669 - subp.py[DEBUG]: Running command ['ip', '--json', 'addr'] with allowed return codes [0] (shell=False, capture=True)
2024-03-07 02:53:41,679 - handlers.py[DEBUG]: finish: init-network/setup-datasource: FAIL: setting up datasource
2024-03-07 02:53:41,680 - util.py[WARNING]: failed stage init
2024-03-07 02:53:41,680 - util.py[DEBUG]: failed stage init
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/cloudinit/cmd/main.py", line 781, in status_wrapper
    ret = functor(name, args)
  File "/usr/lib/python3/dist-packages/cloudinit/cmd/main.py", line 461, in main_init
    init.setup_datasource()
  File "/usr/lib/python3/dist-packages/cloudinit/stages.py", line 539, in setup_datasource
    self.datasource.setup(is_new_instance=self.is_new_instance())
  File "/usr/lib/python3/dist-packages/cloudinit/sources/DataSourceVMware.py", line 187, in setup
    host_info = wait_on_network(self.metadata)
  File "/usr/lib/python3/dist-packages/cloudinit/sources/DataSourceVMware.py", line 965, in wait_on_network
    host_info = get_host_info()
  File "/usr/lib/python3/dist-packages/cloudinit/sources/DataSourceVMware.py", line 888, in get_host_info
    af_inet6.append(convert_to_netifaces_format(addr)) 
  File "/usr/lib/python3/dist-packages/cloudinit/sources/DataSourceVMware.py", line 845, in convert_to_netifaces_format
    "broadcast": addr["bcast"],
KeyError: 'bcast'

File: cloudinit/sources/DataSourceVMware.py (hunk 1 above)
Comment (PengpengSun) [reply cloudinit/sources/DataSourceVMware.py]: Thanks for the review!
I've updated this function as Brett's suggestion as below.

File: tests/unittests/sources/test_vmware.py (hunk 2 above)
Comment (PengpengSun) [reply tests/unittests/sources/test_vmware.py:167]: Added a new test for ipv6 PEER address.

File: cloudinit/sources/DataSourceVMware.py (hunk 3 above)
Comment (PengpengSun) [reply cloudinit/sources/DataSourceVMware.py]: Updated this function and the above convert_to_netifaces_ipv4_format function as suggested, thanks!

**Review by holmanb — APPROVED** [review]
LGTM, thanks @PengpengSun!
